core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -11,53 +11,240 @@
+      long fileId2, long pageIndex2) {
+    if (fileId + pageIndex < fileId2 + pageIndex2) {
+      return new Pair<>(getPageLock(fileId, pageIndex), getPageLock(fileId2, pageIndex2));
+    } else {
+      return new Pair<>(getPageLock(fileId2, pageIndex2), getPageLock(fileId, pageIndex));
+    }
   }
 
   @Override
-  public ReadableByteChannel get(long fileId, long pageId) throws IOException {
-    if (!mMetaStore.hasPage(pageId)) {
+  public void put(long fileId, long pageIndex, byte[] page) throws IOException {
+    long victimFileId = 0;
+    long victimPageIndex = 0;
 
+    ReadWriteLock pageLock = getPageLock(fileId, pageIndex);
+    try (LockResource r = new LockResource(pageLock.writeLock())) {
+      boolean alreadyCached;
+      boolean needEvict = false;
+      try (LockResource r2 = new LockResource(mMetaLock.writeLock())) {
+        alreadyCached = mMetaStore.hasPage(fileId, pageIndex);
+        if (!alreadyCached) {
+          needEvict = mPageStore.size() + 1 > mCacheSize;
+          if (needEvict) {
+            Pair<Long, Long> victim = mEvictor.evict();
+            victimFileId = victim.getFirst();
+            victimPageIndex = victim.getSecond();
+          } else {
+            mMetaStore.addPage(fileId, pageIndex);
+          }
+        }
+      }
+      if (alreadyCached) {
+        try {
+          mPageStore.delete(fileId, pageIndex);
+        } catch (PageNotFoundException e) {
+          // this should never happen with proper locking
+          LOG.error("failed to delete page {} {} from page store", fileId, pageIndex, e);
+        }
+        mEvictor.updateOnPut(fileId, pageIndex);
+        mPageStore.put(fileId, pageIndex, page);
+      } else if (!needEvict) {
+        mEvictor.updateOnPut(fileId, pageIndex);
+        mPageStore.put(fileId, pageIndex, page);
+      }
     }
-    mEvictor.updateOnGet(pageId);
-    return null;
+
+    Pair<ReadWriteLock, ReadWriteLock> pageLockPair =
+        getPageLockPair(fileId, pageIndex, victimFileId, victimPageIndex);
+    try (LockResource r1 = new LockResource(pageLockPair.getFirst().writeLock());
+        LockResource r2 = new LockResource(pageLockPair.getSecond().writeLock())) {
+      try (LockResource r3 = new LockResource(mMetaLock.writeLock())) {
+        if (mMetaStore.hasPage(fileId, pageIndex)) {
+          LOG.warn("fileId {} pageIndex {} is already inserted by a racing thread",
+              fileId, pageIndex);
+          return;
+        }
+        if (!mMetaStore.hasPage(victimFileId, victimPageIndex)) {
+          LOG.warn("fileId {} pageIndex {} is already evicted by a racing thread",
+              fileId, pageIndex);
+          return;
+        }
+        try {
+          mMetaStore.removePage(victimFileId, victimPageIndex);
+        } catch (PageNotFoundException e) {
+          // this should never happen with proper locking
+          LOG.error("failed to remove page {} {} from meta store",
+              victimFileId, victimPageIndex, e);
+        }
```

Comment:
Should we do something in this error case, it doesn't seem safe to proceed?

Reply:
Yeah we can throw `IllegalStateException` instead.